File: Design-Patterns-Using-Java/Interpreter/InterpreterPattern.java
// Abstract Expression
interface AbstractExpression {
    void interpret();
}

// Nonterminal Expression
class NonterminalExpression implements AbstractExpression {
    private AbstractExpression expression;

    public NonterminalExpression(AbstractExpression expression) {
        this.expression = expression;
    }

    @Override
    public void interpret() {
        System.out.println("NonTerminalExpression:interpret");
        expression.interpret();
    }
}

// Terminal Expression
class TerminalExpression implements AbstractExpression {
    @Override
    public void interpret() {
        System.out.println("TerminalExpression:interpret");
    }
}

// Client Code
public class InterpreterPattern {
    public static void main(String[] args) {
        AbstractExpression tree = new NonterminalExpression(new TerminalExpression());
        tree.interpret();
    }
}

/* 
NonTerminalExpression:interpret
TerminalExpression:interpret 
*/

File: Design-Patterns-Using-CSharp/Interpreter/InterpreterPattern.cs
using System;

// Abstract Expression
interface AbstractExpression {
    void Interpret();
}

// Nonterminal Expression
class NonterminalExpression : AbstractExpression {
    private AbstractExpression expression;

    public NonterminalExpression(AbstractExpression expression) {
        this.expression = expression;
    }

    public void Interpret() {
        Console.WriteLine("NonTerminalExpression:Interpret");
        expression.Interpret();
    }
}

// Terminal Expression
class TerminalExpression : AbstractExpression {
    public void Interpret() {
        Console.WriteLine("TerminalExpression:Interpret");
    }
}

// Client Code
class InterpreterPattern {
    static void Main(string[] args) {
        AbstractExpression tree = new NonterminalExpression(new TerminalExpression());
        tree.Interpret();
    }
}

/*
NonTerminalExpression:Interpret
TerminalExpression:Interpret
*/
